### ISSUE TITLE
Standalone PC/Mac Cognito Sync is not working when there is no internet

### ISSUE THREAD

**Issue by LancelotBleu** [aws/aws-sdk-unity#16]
When I remove all internet connections in Unity Editor PC/Mac standalone, I'm receiving errors when trying to log unauthenticated without using any FB.   (Note : I previously logged with FB in the past with this user).

Amazon.CognitoSync.SyncManager.Exceptions.DataStorageException: Failed to list dataset metadata ---> Amazon.Runtime.AmazonServiceException: Could not resolve host: cognito-sync.us-east-1.amazonaws.com; No data record of requested type ---> System.Net.WebException: Could not resolve host: cognito-sync.us-east-1.amazonaws.com; No data record of requested type

Repro

On the cognito sync example, I modified slightly the demo in order to bypass Facebook in case of deconnection.

public class LoginController : MonoBehaviour {

    private bool loggingIn = true;

	public bool IsInternetAvailable
	{
		get
		{
			if (Application.internetReachability == NetworkReachability.NotReachable)
			{
				return false;
			}
			
			if (Application.isMobilePlatform == false) 
			{
				var ipAddress = Network.player.ipAddress.ToString();
				if (ipAddress == "0.0.0.0") return false;
				if (ipAddress == "127.0.0.1") return false;
			}
			
			return true;
		}
	}

    private void Start () 
	{
		if (IsInternetAvailable == true)
		{
			FB.Init(InitCallback);
		}
		else
		{
			loggingIn = false;
		}
    }

I received this error :

WebException: Could not resolve host: cognito-sync.us-east-1.amazonaws.com; No data record of requested type
Rethrow as AmazonServiceException: Could not resolve host: cognito-sync.us-east-1.amazonaws.com; No data record of requested type

Since on mobile devices disconnected mode is working, Is it something that is supported on other platforms?

This is very convenient and faster to test SyncConflict on a computer without having to go on the mobile device itself.  

Thanks

**Comment by albertvaka** [comment]
Hi!

Calling Synchronize will fail if you are not connected to the internet, but your data is still saved in the local storage, so no data is lost and you can synchronize it to the cloud the next time you have an Internet connection. In case you can not get a token from Facebook because you are not connected, it's fine to keep playing "offline" as you already implemented.

About the problem you are facing when testing this in the editor, please note that Unity always reports internet status as "reachable" on the editor, even when it's not connected. You can only rely on this value from Unity when running on an actual mobile device, and this seems to be a Unity limitation. 

From the Unity documentation [1]: "Non-handhelds are considered to always be capable of NetworkReachability.ReachableViaLocalAreaNetwork."

[1] http://docs.unity3d.com/ScriptReference/Application-internetReachability.html

Albert

**Comment by LancelotBleu** [comment]
I managed to fix our sync issues with your informations.   Thanks again!

I probably made a rookie mistake by thinking that "DataSet.OnSyncFailure" was meaning DataSet not working when in fact it just means that there is a network sync failure (but the data set is still ok locally).

For the connection issues with Unity I include this section in my code so it could behave correctly on non-mobile platforms.  I'm not sure if it's a safe way to detect disconnection but it works :

        if (Application.isMobilePlatform == false) 
        {
            var ipAddress = Network.player.ipAddress.ToString();
            if (ipAddress == "0.0.0.0") return false;
            if (ipAddress == "127.0.0.1") return false;
        }

Now I just have a small problem using the DataSet.OnSyncConflict callback.    I noticed that the "SyncConflict.LocalRecord.DeviceLastModifiedDate" is not Utc based (it's region specific) and that "SyncConflict.RemoteRecord.DeviceLastModifiedData" is Utc based.    When I compare the two dates, I could not rely on the date to resolve my conflicts by time.   Is there another way to do that?

Nicolas
